Support `render(output_format = NULL)` default option

plantarum opened the pull request:
added 'default' output to markdown-exporters

When rmarkdown::render is called with the argument `output_format = NULL`
it will process the *first* option specified in the yaml preamble. This
differs from the 'auto detect' option, which processes *all* options
specified in the preamble. If no outputs are defined in the preamble, it
will default to HTML.

vspinu commented:
This looks good. Thanks! But can you check that NULL is the only posibility? Maybe there is a way to pass "DEFAULT" or something instead of NULL, then we could avoid creating a dedicated handler for this case.

plantarum commented:
I agree this would be better. My first attempt looked like this:

```
(defcustom poly-r-markdown-exporter
  (pm-shell-exporter :name "Rmarkdown"
                     :from
                     '(("Rmarkdown"  "\\.[rR]?md\\|rapport\\'" "R Markdown"
                        "Rscript -e \"rmarkdown::render('%i', output_format = %t, output_file = '%o')\""))
                     :to
                     '(("auto" . pm--rmarkdown-shell-auto-selector)
                       ("default" "pdf" "default" "NULL")
                       ("html" "html" "html document" "'html_document'")
                       ...
```

I removed the `'` around `%t` in the render call, and added the single quote to the `t-spec` argument (i.e., `"'html_document'"`). This allowed me to use the unquoted value `"NULL"` in the `t-spec` for `default`. 

This is fine, except that this approach requires you specify the file extension. Which means you have to pre-set it to pdf or html, or you have to provide separate options, which makes for an overly complex user interaction.

I tried to find where polymode applies the file extension, but couldn't track it down. If it's possible to remove the `EXT` argument entirely, and let render (or whatever creates the output) set the extension automatically, then this approach would be simpler than my new handler.

vspinu commented:
@plantarum I actually meant to check, maybe there is a special string value that rmarkdown supports instead of NULL but which would be equivalent to it. I have checked `rmarkdown::render` and it's not there. Maybe it might be a good idea to propose special value like "first" to rmarkdown maintainers? WDYT?

plantarum commented:
I don't have a preference either way. It's possible to pass unquoted (`NULL`) or quoted (`'html_document'`) arguments with the change I suggested above, so it could be handled at polymode without a new "first" option. Or it could be done upstream in `rmarkdown` as you point out. Either way would end up with the same result.

However, this doesn't address the problem of the file extension being imposed by polymode with the current system. Either way (change here or upstream), we'd need polymode to let render set the file extension, without overriding it via the `EXT` parameter in the `pm-shell-exporter` `:to` slot.

vspinu commented:
Doesn't matter much, two extra function is not such a big deal. Thanks!

plantarum commented:
Wonderful, thanks for all your help with these!
